src/compiler/types.ts

```diff
@@ -634,10 +635,14 @@ namespace ts {
         elementTypes: NodeArray<TypeNode>;
     }
 
-    export interface UnionTypeNode extends TypeNode {
+    export interface UnionOrIntersectionTypeNode extends TypeNode {
         types: NodeArray<TypeNode>;
     }
 
+    export interface UnionTypeNode extends UnionOrIntersectionTypeNode { }

```

Comment:
add brands so that these UnionTypeNode  and IntersectionTypeNode are distinguished and not assignable to each other.